File: Math_Refresher_for_Machine_Learning/matlab/ch4.m
% Chapter 4

% Start of script
%-------------------------------------------------------------------------%
close all;                   	% close all figures
clearvars; clearvars -global;	% clear all variables
clc;                         	% clear the command terminal
format shortG;                 	% pick the most compact numeric display
format compact;                	% suppress excess blank lines

% Central Limit Theorem
%-------------------------------------------------------------------------%
% load coin flip
coin_flip

% calculate distribution
n = 1000;
% t = rand(n,1);
trials = 1:n;
heads = 0;
P_heads = ones(size(t));
for ii = 1:n
    if (t(ii) < 0.5)
        heads = heads + 1;
    end
    P_heads(ii) = heads/ii;  
end

% animate plot
figure
for ii = 1:n
    plot(trials(1:ii),P_heads(1:ii))
    ylim([0 1]);
    xlabel('Trial Number')
    ylabel('Probabilty of Heads after n trials')
    title ('Trial Number vs Percent Heads')
    drawnow;
end

%% Ex 4.1
%-------------------------------------------------------------------------%

n = 1:50;

% 1
A_n = 2 - 1./(x.^2);
figure
for ii = n
    plot(n(1:ii),A_n(1:ii))
    ylim([1 2])
    xlabel('Trial Number')
    ylabel('A_n')
    title ('Trial Number vs A_n')
    drawnow;
end

% 2
B_n = (n.^2 + 1)./n;
figure
for ii = n
    plot(n(1:ii),B_n(1:ii))
    ylim([0 60])
    xlabel('Trial Number')
    ylabel('B_n')
    title ('Trial Number vs B_n')
    drawnow;
end

% 3
C_n = (-1).^n .* (1-(1./n));
figure
for ii = n
    plot(n(1:ii),C_n(1:ii))
    ylim([-1 1])
    xlabel('Trial Number')
    ylabel('C_n')
    title ('Trial Number vs C_n')
    drawnow;
end


%% Fig 4.3
%-------------------------------------------------------------------------%

x = -50:1:50;

% f(x) = sqrt(x)
f_x = real(sqrt(x));
figure
for ii = 1:length(x)
    plot(x(1:ii),f_x(1:ii))
    ylim([0 10])
    xlabel('Trial Number')
    ylabel('f(x)')
    title ('Trial Number vs f(x)')
    drawnow;
end

% f(x) = 1/x
f_x = 1 ./ x;
figure
for ii = 1:length(x)
    plot(x(1:ii),f_x(1:ii))
    ylim([-1 1])
    xlabel('Trial Number')
    ylabel('f(x)')
    title ('Trial Number vs f(x)')
    drawnow;
end


%% Ex 4.4
%-------------------------------------------------------------------------%

% f(x) = sqrt(x)
x = 1:50;
f_x = sqrt(x);
figure
for ii = 1:length(x)
    plot(x(1:ii),f_x(1:ii))
    ylim([0 10])
    xlabel('Trial Number')
    ylabel('f(x)')
    title ('Trial Number vs f(x) = sqrt(x)')
    drawnow;
end


% f(x) = e^x
x = -5:1:5;
f_x = exp(x);
figure
for ii = 1:length(x)
    plot(x(1:ii),f_x(1:ii))
    ylim([0 150])
    xlabel('Trial Number')
    ylabel('f(x)')
    title ('Trial Number vs f(x) = e^x')
    drawnow;
end


% f(x) = 1+(1/x^2)
x = -2:0.05:2;
f_x = 1+(1 ./x.^2);
figure
for ii = 1:length(x)
    plot(x(1:ii),f_x(1:ii))
    ylim([0 150])
    xlabel('Trial Number')
    ylabel('f(x)')
    title ('Trial Number vs f(x) = 1+(1/x^2)')
    drawnow;
end


% f(x) = floor(x)
x = 0:0.05:5;
f_x = floor(x);
figure
for ii = 1:length(x)
    plot(x(1:ii),f_x(1:ii),'*')
    ylim([0 5])
    xlabel('Trial Number')
    ylabel('f(x)')
    title ('Trial Number vs f(x) = floor(x)')
    drawnow;
end

File: Books/Math_Refresher/python/ch4.py
# ch4.py
import numpy as np
import matplotlib.pyplot as plt
from coin_flip import t

plt.close('all')

# Central Limit Theorem section (same as before)
n = 1000
trials = np.arange(1, n+1)
heads = 0
P_heads = np.ones_like(t, dtype=float)
for ii in range(n):
    if t[ii] < 0.5:
        heads += 1
    P_heads[ii] = heads / (ii + 1)
plt.figure()
for ii in range(n):
    plt.clf()
    plt.plot(trials[:ii+1], P_heads[:ii+1])
    plt.ylim([0, 1])
    plt.xlabel('Trial Number')
    plt.ylabel('Probability of Heads after n trials')
    plt.title('Trial Number vs Percent Heads')
    plt.pause(0.001)

# Ex 4.1: Animated plots for sequences

n_vals = np.arange(1, 51)

# 1. A_n = 2 - 1/(n^2)
A_n = 2 - 1/(n_vals**2)
plt.figure()
for ii in range(len(n_vals)):
    plt.clf()
    plt.plot(n_vals[:ii+1], A_n[:ii+1])
    plt.ylim([1, 2])
    plt.xlabel('Trial Number')
    plt.ylabel('A_n')
    plt.title('Trial Number vs A_n')
    plt.pause(0.1)

# 2. B_n = (n^2 + 1)/n
B_n = (n_vals**2 + 1) / n_vals
plt.figure()
for ii in range(len(n_vals)):
    plt.clf()
    plt.plot(n_vals[:ii+1], B_n[:ii+1])
    plt.ylim([0, 60])
    plt.xlabel('Trial Number')
    plt.ylabel('B_n')
    plt.title('Trial Number vs B_n')
    plt.pause(0.1)

# 3. C_n = (-1)^n * (1 - (1/n))
C_n = ((-1)**n_vals) * (1 - 1/n_vals)
plt.figure()
for ii in range(len(n_vals)):
    plt.clf()
    plt.plot(n_vals[:ii+1], C_n[:ii+1])
    plt.ylim([-1, 1])
    plt.xlabel('Trial Number')
    plt.ylabel('C_n')
    plt.title('Trial Number vs C_n')
    plt.pause(0.1)

# Fig 4.3: Animated plots for functions

x = np.arange(-50, 51, 1)
# f(x) = sqrt(x) (only for x>=0; for x<0, take 0)
f_x = np.sqrt(np.maximum(x, 0))
plt.figure()
for ii in range(len(x)):
    plt.clf()
    plt.plot(x[:ii+1], f_x[:ii+1])
    plt.ylim([0, 10])
    plt.xlabel('Trial Number')
    plt.ylabel('f(x)')
    plt.title('Trial Number vs f(x)')
    plt.pause(0.05)

# f(x) = 1/x (avoid division by zero)
f_x = 1 / x.astype(float)
f_x[np.isinf(f_x)] = np.nan
plt.figure()
for ii in range(len(x)):
    plt.clf()
    plt.plot(x[:ii+1], f_x[:ii+1])
    plt.ylim([-1, 1])
    plt.xlabel('Trial Number')
    plt.ylabel('f(x)')
    plt.title('Trial Number vs f(x)')
    plt.pause(0.05)

# Ex 4.4: f(x) = sqrt(x)
x = np.arange(1, 51)
f_x = np.sqrt(x)
plt.figure()
for ii in range(len(x)):
    plt.clf()
    plt.plot(x[:ii+1], f_x[:ii+1])
    plt.ylim([0, 10])
    plt.xlabel('Trial Number')
    plt.ylabel('f(x)')
    plt.title('Trial Number vs f(x) = sqrt(x)')
    plt.pause(0.05)

# f(x) = e^x
x = np.arange(-5, 6, 1)
f_x = np.exp(x)
plt.figure()
for ii in range(len(x)):
    plt.clf()
    plt.plot(x[:ii+1], f_x[:ii+1])
    plt.ylim([0, 150])
    plt.xlabel('Trial Number')
    plt.ylabel('f(x)')
    plt.title('Trial Number vs f(x) = e^x')
    plt.pause(0.05)

# f(x) = 1 + (1/x^2) (be cautious at x=0)
x = np.arange(-2, 2.05, 0.05)
f_x = 1 + 1/(x**2)
plt.figure()
for ii in range(len(x)):
    plt.clf()
    plt.plot(x[:ii+1], f_x[:ii+1])
    plt.ylim([0, 150])
    plt.xlabel('Trial Number')
    plt.ylabel('f(x)')
    plt.title('Trial Number vs f(x) = 1+(1/x^2)')
    plt.pause(0.05)

# f(x) = floor(x)
x = np.arange(0, 6, 0.1)
f_x = np.floor(x)
plt.figure()
for ii in range(len(x)):
    plt.clf()
    plt.plot(x[:ii+1], f_x[:ii+1], '*')
    plt.ylim([0, 5])
    plt.xlabel('Trial Number')
    plt.ylabel('f(x)')
    plt.title('Trial Number vs f(x) = floor(x)')
    plt.pause(0.05)

plt.show()
